### PR TITLE
Add TokenCaseIterator

### PR DESCRIPTION
**Pull request by mbeccati**

### CODE REVIEW

**Comment by mbeccati** [comment]
@keradus during a refactoring, I came up with this idea to reuse the same logic in Token::equalsAny and Tokens::findSequence. Thoughts?

**Comment by keradus** [comment]
Iteration idea could be helpfull.
Is this implementation already done ?

**Comment by mbeccati** [comment]
@keradus yes, I didn't merge it as it could be considered a bit too much

**Comment by keradus** [comment]
Then I will write down my thoughts here,

File: Symfony/CS/Tokenizer/TokenCaseIterator.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+     * Class constructor
+     *
+     * @param array      $tokens        array of tokens or token prototypes
+     * @param bool|array $caseSensitive global case sensitiveness or an array of booleans, whose keys should match
+     *                                  the ones used in $others. If any is missing, the default case-sensitive
+     *                                  comparison is used.
+     */
+    public function __construct(array $tokens, $caseSensitive)
+    {
+        $this->tokenIterator = new \ArrayIterator($tokens);
```
Comment (keradus) [inline Symfony/CS/Tokenizer/TokenCaseIterator.php]: So, you create `TokenCaseIterator`, that implement `SeekableIterator`, only to aggregate `ArrayIterator`, which implements... `SeekableIterator`...
oh my...

**Comment by mbeccati** [comment]
@keradus Any better?

**Comment by keradus** [comment]
Better, but...



before:
```php
public function equalsAny(array $others, $caseSensitive = true)
{
    foreach ($others as $key => $other) {
        $cs = is_array($caseSensitive) ? $cs = isset($caseSensitive[$key]) ? $caseSensitive[$key] : true) : $caseSensitive;
        if ($this->equals($other, $cs)) {
            return true;
        }
    }
    return false;
}
```

or even:
```php
public function equalsAny(array $others, $caseSensitive = true)
{
    $caseSensitiveDefault = is_array($caseSensitive) ? true : $caseSensitive;
    if (!is_array($caseSensitive)) {
        $caseSensitive = array();
    }

    foreach ($others as $key => $other) {
        if ($this->equals($other, isset($caseSensitive[$key]) ? $caseSensitive[$key] : $caseSensitiveDefault)) {
            return true;
        }
    }
    return false;
}
```

proposed:
```php
public function equalsAny(array $others, $caseSensitive = true)
{
    foreach (new TokenCaseIterator($others, $caseSensitive) as $other) {
        if ($this->equals($other['token'], $other['caseSensitive'])) {
            return true;
        }
    }
}
```

But, the proposed solution add one (uff... not two anymore) abstract layer. For now we have `Tokens`, which implements `SplFixedArray`, which implements `Iterator`. You created extra layer `TokenCaseIterator` which implemented `Iterator`. You need sth that implements Iterator but you already have one.
And whole that extra layer just to move one line to other place. And that new iterator have ugly interface - returning array of `token` and `caseSensitive`, but hey, we gave both on input, just to take it back on output.
For me it's triumph of form over content.

**Comment by mbeccati** [comment]
I get your point. This was a way to reuse the same logic in equalsAny and findSequence without duplicating the code in multiple places. Both are expected to accept either a bool or an array as $caseSensitive and the Iterator simply does that.

It would be great to reuse Token and/or Tokens wihtout a new Iterator, but I don't see any good way to add case-sensitivity to either/both of them while keeping BC with the array prototypes.

**Comment by keradus** [comment]
What about array prototypes? You didn't change the handling of them in this PR

**Comment by keradus** [comment]
Your 'same' logic is one line, and if you want you can place it (change `bool` into `bool[]` when needed) in helper method

**Comment by mbeccati** [comment]
I meant that ideally case-sensitiveness should be part of Token and its prototype, e.g. 
```php
array(T_STRING, 'functionName', $caseSensitive = false)
```
I don't like using two arrays, but sometimes BC doesn't help designing something clean.

I'm not sure I got what you mean by your last comment (change bool into bool[])

**Comment by keradus** [comment]
> I meant that ideally case-sensitiveness should be part of Token and its prototype

No it does not. case sensitiveness is no part of info about token. It's just sometimes usefull when you filter array of tokens, but it is not a part of single token. Token itself doesn't care if you filter using cs or not. Keep single responsibility.

------------------


> I'm not sure I got what you mean by your last comment (change bool into bool[])

So...
>  I came up with this idea to reuse the same logic in Token::equalsAny and Tokens::findSequence.

That same logic you need to use twice is converting `bool|array $caseSensitive` parameter to array of bools (`bool[]`) to pass for `equals`. That logic is whole one line:
`$cs = is_array($caseSensitive) ? $cs = isset($caseSensitive[$key]) ? $caseSensitive[$key] : true) : $caseSensitive;`
If you don't wan't to repeat that in to method you mentioned simply move that into separate method.

**Comment by mbeccati** [comment]
Ok, makes sense.

**Pull request closed without merging**